Add first version of code insights creation UI

vovakulikov opened the pull request:
This PR is the first out of three PR for creation UI. 
1. **This Pull Request.**
2. [Introduction page for creation UI](https://github.com/sourcegraph/sourcegraph/pull/20482). https://github.com/sourcegraph/sourcegraph/pull/20482
3. [Creation UI for lang stats insights](https://github.com/sourcegraph/sourcegraph/pull/20548)

This PR is about to add the first implementation of code insight [creation UI.](https://www.figma.com/file/QfeID0EEYMEPdEv8pbGULB/Code-Insights-WIP?node-id=262%3A11)

Compare to the existing approach users can go to the creation page and create code insight by filling form inputs and therefore modify the user or organization settings (the existing approach assumes that the user has to edit that settings configuration manually). You can see the creation form for search-code-insight on a screenshot below.

<img width="988" alt="image" src="https://user-images.githubusercontent.com/18492575/116142873-17ae0680-a6e3-11eb-91c7-20f3b7a1c0a1.png">

**How to test** 
1. Open dev tools and set **_enableCodeInsightCreationUI_** field in local storage to `true`.
2. Go to `/insights` and click configure insight.
3. You must see `/insights/create` page with insight creation form.

**Tech and UX details** 
- For form validation and form management these changes use [react-final-form-hooks](https://github.com/final-form/react-final-form-hooks)
- In case if a user tried to submit the form with invalid inputs we focus first on invalid input and by that force to fill all invalid parts of the creation form. 
- All styles for creation form were written in CSS modules approach. 

**Existing UX problems which will be solved in next iterations** 
- Poor repositories input 
- Support case when a user was included in two or more organizations (add select or some part of UI)

vovakulikov commented:
I'm about to write some unit tests to cover core functionality of creation form, but beside that this PR is ready for review.

eseliger commented on client/sandboxes/code-insights/webpack.config.js:
```diff
@@ -24,8 +24,26 @@ module.exports = {
+              localsConvention: 'camelCase',
+              modules: {
+                localIdentName: '[name]__[local]_[hash:base64:5]',
+              },
```
I think this config doesn't work with our version of css loader, see here: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/webpack.config.js#L200-206 #18584 ( I assume you copied that before I merged the css loader v4 upgrade)

vovakulikov replied:
You right. I've fixed that already. I took this config from my pet project config, so maybe I should just use our web configuration file from web dir that you suggested in PR about code insights sandbox.

eseliger commented on client/web/src/enterprise/batches/preview/BatchChangePreviewStatsBar.scss:
```diff
@@ -1,3 +1,5 @@
+//@import "~bootstrap/scss/mixins/breakpoints";
+
```
What does this do?

vovakulikov replied:
Sorry. That's leak from other branch. I'll remove this of course. Thanks for catching this

eseliger commented on client/web/src/insights/InsightsRouter.tsx:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Feature flag for new code insights creation UI.
+ * */
+const isCreationUIEnabled = localStorage.getItem('enableCodeInsightCreationUI')
```
maybe use the `useLocalStorage` hook, so it's not only evaluated at page load?

limitedmage replied:
Can we use the experimentalFeatures setting instead of localStorage for this feature flag?

vovakulikov replied:
@eseliger thanks. I didn't notice that we have the special hook for that. 

@limitedmage  That was my original thought but this functionality is not ready (completed) to share with customers, since experimentalFeatures available in user/org settings I would't do it. Plus we already have experimentalFeatures for codeInsights itself so experimentalFeatures of experimentalFeatures. For these changes, I picked just local storage because I still want to share this thing but without experimental/public release.

eseliger reviewed (commented):
Haven't read all of it yet, but heading into meetings soon so leaving the first portion of comments now :)

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-field/FormField.module.scss:
```diff
@@ -0,0 +1,18 @@
+
+    &__error {
+        color: var(--danger);
+        font-size: 0.875em;
```
You can achieve this same style with `<small class="text-danger">`. Especially the font size we should very rarely should override to make sure we stick to the typography scale (and adapt to the redesign).

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-field/FormField.tsx:
```diff
@@ -0,0 +1,53 @@
+                ref={useMergeRefs([localInputReference, reference])}
+            />
+
+            {error && <span className={styles.formFieldError}>*{error}</span>}
```
We don't have this asterisk in other forms, can we remove that?
```suggestion
            {error && <span className={styles.formFieldError}>{error}</span>}
```

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-color-input/FormColorInput.module.scss:
```diff
@@ -0,0 +1,79 @@
+        margin: 0;
+    }
+
+    &__colour-content {
```
**nit**: we usually use American English, which is also how CSS spells `color`.
```suggestion
    &__color-content {
```

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-series-input/FormSeriesInput.tsx:
```diff
@@ -0,0 +1,192 @@
+                            'button'
+                        )}
+                    >
+                        Cancel
```
I think "Discard" is a bit more fitting as this is more of a draft you're throwing away, as opposed to an ongoing operation you want to abort/cancel.
```suggestion
                        Discard
```
@quinnkeast do we have content guidelines on this?

quinnkeast replied:
I totally understand the argument here... however, "Cancel" is still a logical opposition to the action being noped-out of, which in this case, is "**Create** / **Cancel**," and in the future, "**Edit** or **Save changes** / Cancel". I think consistency in the action label here works well. 

(Whatever changes we do here, would logically then apply to code monitoring and search contexts in the future.)

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-series-input/FormSeriesInput.tsx:
```diff
@@ -0,0 +1,192 @@
+            <InputField
+                title="Query"
+                placeholder="ex. spatternType:regexp const\\s\\w+:\\s(React\\.)?FunctionComponent"
+                description="Do not include the repo: filter as it will be added automatically for the current repository"
```
The design had the `repo:` filter wrapped in `<code>`, to make sure the colon is not read like a grammatical colon.

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-series-input/FormSeriesInput.tsx:
```diff
@@ -0,0 +1,192 @@
+
+            <InputField
+                title="Query"
+                placeholder="ex. spatternType:regexp const\\s\\w+:\\s(React\\.)?FunctionComponent"
```
Typo
```suggestion
                placeholder="ex. patternType:regexp const\\s\\w+:\\s(React\\.)?FunctionComponent"
```

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-series-input/FormSeriesInput.tsx:
```diff
@@ -0,0 +1,192 @@
+        <div className={classnames(styles.formSeriesInput, className)}>
+            <InputField
+                title="Name"
+                placeholder="ex. Function component"
```
According to our [content guidelines](https://about.sourcegraph.com/handbook/communication/content_guidelines/style_and_mechanics#abbreviations-acronyms-latinisms-jargon), we should use complete English words and avoid abbreviations
```suggestion
                placeholder="Example: Function components"
```

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-series/FormSeries.module.scss:
```diff
@@ -0,0 +1,62 @@
+        margin-top: 0.5rem;
+    }
+
+    &__add-button {
+        width: 100%;
+        padding: 1rem;
+        color: var(--link-color);
+        background-color: transparent;
+        border: none;
+        transition: background-color 0.2s ease;
+
+        &:hover {
+            background-color: var(--link-hover-bg-color);
+        }
+    }
```
This class seems to recreate the standard `btn-link` class. See e.g. [our "Buttons" story in Storybook](https://storybook.sgdev.org/?path=/story/branded-global-styles--buttons)

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-series/FormSeries.module.scss:
```diff
@@ -0,0 +1,62 @@
+    display: flex;
+    flex-direction: column;
+
+    &__input {
+        border-radius: 4px;
+        border: 1px solid var(--border-color);
+        padding: 1rem;
+    }
```
This class seems to recreate `form-control`. See e.g. in [our "Forms" story in Storybook](https://storybook.sgdev.org/?path=/story/branded-global-styles--forms)

vovakulikov replied:
These styles are needed for the series input/block but this is not the input element in general way.

<img width="506" alt="image" src="https://user-images.githubusercontent.com/18492575/116550203-2113ba80-a8ff-11eb-8f96-6b7006102cbd.png">

felixfbecker replied:
I think that's a `card`?

felixfbecker commented:
@vovakulikov it would be helpful to have stories in Storybook for this PR (not just the custom  playground). The reason is that those will get screenshotted in Chromatic, and it would make reviewing a lot easier because the Storybook can be accessed directly for this branch without running locally (which is not possible for the custom playground). Storybook then also makes it easy to toggle different themes (including the redesign) thanks to the tooling that the frontend platform team added for it.

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/CreationSearchInsightPage.tsx:
```diff
@@ -0,0 +1,117 @@
+
+            try {
+                const settings = await getSubjectSettings(subjectID).toPromise()
+                const content = JSON.parse(settings.contents) as object
```
**issue**: this won't work, because settings are JSONC and can contain comments and trailing commas.

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/CreationSearchInsightPage.tsx:
```diff
@@ -0,0 +1,117 @@
+                await updateSubjectSettings(
+                    platformContext,
+                    subjectID,
+                    JSON.stringify(newSettings, null, 2)
```
**issue**: Parsing and serializing the JSON like this will discard all comments and formatting. We have code in the webapp to modify JSONC preserving comments and formatting - check e.g. the settings quick actions on the settings editor.

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/CreationSearchInsightPage.tsx:
```diff
@@ -0,0 +1,117 @@
+            try {
+                const settings = await getSubjectSettings(subjectID).toPromise()
+                const content = JSON.parse(settings.contents) as object
+                const insightID = uuid.v4()
```
Instead of a UUID, can we use `camelCase(title)` so that the settings remain "hand-editable"?

vovakulikov replied:
Resolve this comment in favour of https://github.com/sourcegraph/sourcegraph/pull/20032#discussion_r621448175

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/CreationSearchInsightPage.tsx:
```diff
@@ -0,0 +1,117 @@
+                        repositories: values.repositories.split(','),
+                        series: values.series.map(line => ({
+                            name: line.name,
+                            // Query field is a reg exp field for code insight query setting
+                            // Native html input element adds escape symbols by himself
+                            // to prevent this behavior below we replace double escaping
+                            // with just one series of escape characters e.g. - //
```
Are you sure about this?  `HTMLInputElement.value` should be the exact value as entered by the user, not escaped in any way afaict. Could it be that there is a bug elsewhere?

vovakulikov replied:
That was surprise to me as well. It seems like input should return exact same value and if you try to print this in console.log you will get value without double escaping symbols but if you stop evaluation by breakpoints you will see that value has this escaping thing. I don't know why actually this works like this but I've faced a problem when I tried do a serialisation to json with this value. I will try to create a minimal demo to demonstrate this maybe I'm missing here something

felixfbecker replied:
Keep in mind that within _JSON_ it _has_ to be escaped (it's expected that `JSON.stringify` would do that, and it would break if it didn't).

E.g. `test\(` would need to become `"test\\("` because `\` is a reserved character in JSON strings.

vovakulikov replied:
Minimal viable example of this problem 
https://codesandbox.io/s/distracted-wind-ozlk8?file=/src/App.js

```
First render
[Log] {value: "", json: "{\"value\":\"\"}"}

Typed \ symbol
[Log] {value: "\\", json: "{\"value\":\"\\\\\"}"} ```

felixfbecker replied:
^ in that example the extra backslash is added by `JSON.stringify()`. As I mentioned above,  that's expected. If you were to parse that back into a string with `JSON.parse()`, the extra backslash would be gone again.
I think this extra handling here actually introduces a bug and it would function correctly if removed 🤔

vovakulikov replied:
Now we don't have JSON.stringify calls in a creation UI 
If I console.log the value of series with some query with backslash I'm getting doubled backslashes 

<img width="506" alt="image" src="https://user-images.githubusercontent.com/18492575/118042396-afb31d80-b37c-11eb-8d1c-0349a87847b3.png">

felixfbecker reviewed (commented):
High-level feedback:
- It seems like a lot of custom form components are added here that are not specifically related to insights, but are in the insights folder. There is also an additional dependency added for forms. I'd challenge you: Do we need these? All of the forms in our app currently as far as I'm aware get by with our simple `<Form>` component, native `<input>`s (+`form-control` CSS class), CSS utilities, native HTML attributes for validation like `required`, and React `useState()` for state. Comparatively, the component structure here feels a bit over-engineered (it may be that I'm missing a valid reason why this is different though, in which case please point it out). For more complex/custom validation, we also have a `useInputValidation()` hook that plugs into the native constraint validation API. Since inconsistency is a challenge that has been raised for our codebase, I'd be hesitant to add _yet another_ new way of doing things, especially if it's isolated for code insights. I'd see if we can follow those existing patterns (sticking closely to classes we already have and platform features). If that is not possible, it would be great if you could speak to the special requirements here and we can involve the frontend platform team on an approach that can be standardized on.
- There is a lot of CSS that intuitively feels like it could be reduced heavily (like ~10% of what's currently used) by using existing classes.
- The radio buttons in the screenshot look wrong/too large. Can we just use the native browser radio buttons? That's what we use elsewhere, and how the designs should generally be interpreted. See [this note in Figma](https://www.figma.com/file/NIsN34NH7lPu04olBzddTw/Design-Refresh-Systemization-source-of-truth?node-id=908%3A4001).

vovakulikov commented on client/web/src/insights/core/backend/insights-api.ts:
```diff
@@ -57,4 +58,12 @@ export class InsightsAPI implements ApiService {
             )
         )
+
+    public getSubjectSettings = (id: string): Observable<SubjectSettingsResult> =>
```
Main reason why I came up with classes for backend API is you can easily create a mock for testing purposes. 
If you go to demo component https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/sandboxes/code-insights/src/demo.tsx#L28 Here we mock all API methods by one line which I've found useful.

But I think I can remove class here and use some function factory like createInsightAPI or something like this.

felixfbecker replied:
Yeah, I think it would be possible to define a simple object that satisfies the interface for tests. See e.g. definitions of `commonProps` in stories that are spread out.

vovakulikov commented on client/web/src/insights/pages/create/components/create-insight-form/CreateInsightForm.tsx:
```diff
@@ -0,0 +1,240 @@
+
+    return (
+        // eslint-disable-next-line react/forbid-elements
+        <form onSubmit={handleSubmit} className={classnames(className, styles.createInsightForm)}>
```
Yes, we can use here our own form component, since I implemented validation by myself here I didn't add this form, but I see Form component has native form API so it wouldn't be a problem replace form to Form

vovakulikov commented on client/web/src/insights/InsightsRouter.tsx:
```diff
@@ -0,0 +1,64 @@
+                exact={true}
+            />
+
+            {isCreationUIEnabled && (
```
Yes, But since this is just a first implementation of some part of creation UI flow I put this behind some local storage flag (plus I have to reconsider route paths once again when we have full picture of creation UI ). As soon as we have all parts of creation UI (lang stats insight form and intro creation UI page I'll remove this local storage flag)

vovakulikov commented on client/web/src/insights/pages/create/CreateInsightPage.tsx:
```diff
@@ -0,0 +1,114 @@
+
+            try {
+                const settings = await getSubjectSettings(subjectID).toPromise()
+                const content = JSON.parse(settings.contents) as object
```
Nice catch. Should I use here just string concatenation then or we have some custom parser for it?

eseliger replied:
`@sqs/jsonc-parser`

vovakulikov replied:
See your comment below about that. Got it

vovakulikov commented:
@felixfbecker I agree with all points from you maybe except the part about validation. 

First thing if be honest I didn't notice that we have a validation hook. 

So my point here it seems like our own hook `useInputValidation` can work only with native inputs which are good but sometimes (creation UI case) this is not enough in API flexibility. I'll explain what I mean. Sometimes we want to have some high-level abstraction for some complex part of the form. In our case, this is a series input that actually looks like a sub-form but acts as an input for the top-level form component. 

So in the case of the component with these lightweight hooks from 'react-final-form' we don't need to think about API much we only have to implement value prop and onChange callback and we are good. All validation and UX logic will be working. But disadvantage here is we do not use standard HTML validation API. 

The second thing here is the UX aspects. 
Classic form validation requires not showing error until some input is touched and blurred and immediately hides when the user is providing a correct value for the input. So all these things about touched, focused, and others things are included in these hooks from 'react-final-form'. 

Another thing here if the user tried to submit the form with invalid inputs we should focus first on invalid input and show all input errors. So all these things we have to implement ourselves or extend our existing hook which seems a bit complicated but in the future, we can back to this and reconsider public API and features which this hook provides. (I believe some things our own hook can do already)

Here is interesting article about some UX thing related to form validation 
https://www.nngroup.com/articles/errors-forms-design-guidelines/

felixfbecker commented:
There was one thing I forgot to comment (sorry!), which I want to bring up before addressing your comment because it is somewhat relevant:

> Existing UX problems which will be solved in next iterations
> - A11y. This form built by native inputs and elements so I don't see big problems now (especially for first iteration) but we have to back to this point later anyway because support accessible validation logic and focus management this is something which we need improve here.

When we split work into smaller parts, imo it should never be on the axis of PR 1: "make everything work without accessibility", then PR 2: "make everything accessible". The right axis to chunk work on is different parts and features of the UI, but each feature should always be built with accessibility in mind from the start. This is more than a committment to accessibility, it actually has a very practical reason too: The cost to write accessible code from the start is often not higher than writing it in a non-accessible way. If it's written in an inaccessible way first though, entirely different design decisions may be made for the code, and it's a much larger cost to rewrite everything afterwards to make it accessible. For example, the number 1 easiest way to get most accessibility benefits is to use semantic HTML features (as our [dev docs](https://docs.sourcegraph.com/dev/background-information/web/accessibility#:~:text=this-,Write%20semantic%20HTML,Use,-labels) also say). It is not more costly to chose `<button>` over `<div>`, or the `required` attribute over custom JS for validation, for an initial implementation. But going back _afterwards_ and refactoring everything for accessibility is way more complex than simply sticking with semantic HTML from the start, which means more work and an adverse incentive to not do so.

In this context, "But disadvantage here is we do not use standard HTML validation API." is a disadvantage we shouldn't take lightly. Exposing the requirements and validity of an input field to assistive technologies _automatically_ is a **big** benefit. If we can do that from the start, instead of first writing the whole form in a way that doesn't use semantic attributes, and then later go in to rewrite everything to make it accessible (e.g. either add manual handling with ARIA everywhere, or rewrite everything to use semantic attributes instead), it saves a lot of work.

Regarding the UXs aspects, these are all great considerations, I agree! That "best practice" behavior is actually done by the browser exactly as you describe - and without needing any logic with refs like in this PR, i.e. less code. When you try to submit a form that is _invalid_ in the browser (as in, constraints are not fulfilled, or validity has been set through the validation API) the browser will automatically focus the first invalid input. See https://pauljadam.com/guides/html5-form.html for a demo.

The "validate on blur" behavior is also very easy to implement I think (untested code):
```tsx
const reportValidity = event => event.target.reportValidity()

<input onBlur={reportValidity} />
```
if we agree that this is the behavior we want for our forms, we can even very easily abstract that in our `<Form>` component so that it works for all inputs without any extra code (could also be controlled with a boolean prop):
```tsx
<form onFocusOut={reportValidity}> // focusout instead of blur because it bubbles
```
(there may be some things to iron out here, but I believe it would not be more complex than the current solution, and benefit all of our forms and users of all abilities).

So _I think_ the (very valid!) UX concerns are all covered.

What I'm not clear about yet is this part of your comment:

> but sometimes (creation UI case) this is not enough in API flexibility. I'll explain what I mean. Sometimes we want to have some high-level abstraction for some complex part of the form. In our case, this is a series input that actually looks like a sub-form but acts as an input for the top-level form component.

Can you explain a bit more on how the "subform" complicates this, or prevents us from using the native APIs? (totally possible that there is a legit reason here and I just don't understand it yet)

vovakulikov commented:
@felixfbecker The creation form actually is accessible enough now, We have keyboard navigation and and correct focus management just because we use standard and semantic elements. So I believe only voice over support is the thing which I should check once again regarding these concerns about native validation. 

About this moment with sub-form - According to our Figma designs we have some part of form (sub-form actually) with three fields to create another insight series (screenshot below)

<img width="666" alt="image" src="https://user-images.githubusercontent.com/18492575/116463301-36430780-a873-11eb-890e-653d262796b5.png">

So these series inputs are kind of form component but for creation UI form (top-level component) this series form is just an input `series: []`.

**Top level form** 
 - title - `input`
 - repositories - `input`
 - visibility - `input`
 - **series** : [{ name, query, color}, {...}, ....] - `input` for top level but `kind of form` under the hood
     Internally series is a form with three field with his own validation trigger by button "Done"
     
So this lead us to situation when we must have two or more separate validation system. Example: user created two insights series (screenshot below)

<img width="721" alt="image" src="https://user-images.githubusercontent.com/18492575/116464424-a43bfe80-a874-11eb-835b-976fb608710f.png">

And clicked on first one to edit, we rendered form with pre-filled inputs with series data. 

<img width="676" alt="image" src="https://user-images.githubusercontent.com/18492575/116464582-ccc3f880-a874-11eb-92ad-b921b80e12d5.png">

Then he edited some fields (remove name value) without accepting these changes by done button. 
Next move he clicked on create insight submit button (top level form) he will get some validation error because for native form validation all inputs within form element must be valid - but this is not the case - series input is still valid we didn't commit the user changes so we still have two valid series in series input. 

We have to treat series as input with some value but not as some part of top level form with inputs which need to be validated on top level. With native validation I don't see the way to add some 'stacking'.

vovakulikov commented:
And one moment about this 
>Comparatively, the component structure here feels a bit over-engineered

It's actually not so many components. `InputField` or `FormRadioInput` these are just a small components which encapsulated some layouts according to our creation UI designs (description, errors, title). Same thing for `FormSeries` components. Just because we have not so simple logic behind this form (edit series, add series, remove series) that may look like overhead. I'll try to simplify some parts.

vovakulikov commented:
Maybe I biased here about native validation, but this is just because I'm not sure if we start implementing all these peculiarities about touched - valid/invalid input states we probably would ended up more or less with the same solution like react-final-form or [react-form-hook](https://github.com/react-hook-form/react-hook-form). (last one seems like some compromise here because they use native constraint-based validation API)

vovakulikov commented:
And thank you @felixfbecker for dive deep and really comprehensive and valid comments here about validation. I really appreciate it 🙂

vovakulikov commented on client/web/src/insights/core/backend/types.ts:
```diff
@@ -14,11 +15,18 @@ export interface ViewInsightProviderResult extends ViewProviderResult {
     source: ViewInsightProviderSourceType
 }
 
+export interface SubjectSettingsResult {
```
Generated types for this shape didn't work for me well, because If I couldn't find latest settings I return now `{ id: null }` and then in consumer I handle this case. In generated type we have `Maybe<{ id: number; contents: string }>` which doesn't imply that id might be null.

But maybe I should handle case when entire latestSetting is undefined. need to think about this.

felixfbecker replied:
If you can't find the latest settings, wouldn't `contents` be `null` as well? 🤔

vovakulikov commented on client/web/src/insights/pages/create/components/create-insight-form/CreateInsightForm.module.scss:
```diff
@@ -0,0 +1,56 @@
+.create-insight-form {
```
I went through all styles of creation UI and removed a lot of thing and replaced them with bootstrap classes.

vovakulikov commented on client/web/src/insights/pages/create/CreateInsightPage.tsx:
```diff
@@ -0,0 +1,114 @@
+            try {
+                const settings = await getSubjectSettings(subjectID).toPromise()
+                const content = JSON.parse(settings.contents) as object
+                const insightID = uuid.v4()
```
Yeah. At some point we agreed to have validation that user couldn't create two or more insights with the same title/name field. And this makes sense for user visibility but then I thought this is kind of inconvenient for user who want to share his insights with organisation and he can't because insight with the same name is already created by someone in org. 

Another moment about title validation. There is still some chance that there will be a name collision since if one user submits the form at a time when the first user has already filled in the title but has not yet submitted the form, so he passed validation but he has same title with insight which has been created by another users. 

WDYT @felixfbecker

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-color-input/FormColorInput.tsx:
```diff
@@ -0,0 +1,88 @@
+    className?: string
+}
+
+const DEFAULT_COLOURS = [
+    { color: openColor.red[7], name: 'Red' },
+    { color: openColor.pink[7], name: 'Pink' },
+    { color: openColor.grape[7], name: 'Grape' },
+    { color: openColor.violet[7], name: 'Violet' },
+    { color: openColor.indigo[7], name: 'Indigo' },
+    { color: openColor.blue[7], name: 'Blue' },
+    { color: openColor.cyan[7], name: 'Cyan' },
+    { color: openColor.teal[7], name: 'Teal' },
+    { color: openColor.green[7], name: 'Green' },
+    { color: openColor.lime[7], name: 'Lime' },
+    { color: openColor.yellow[7], name: 'Yellow' },
+    { color: openColor.orange[7], name: 'Orange' },
+]
```
**suggestion**: 
```suggestion
const DEFAULT_COLOURS = Object.keys(openColor)
    .filter(name => name !== 'white' && name !== 'black')
    .map(name => ({ name: startCase(name), color: `var(--oc-${name}-7)` }))
```

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-group/FormGroup.module.scss:
```diff
@@ -0,0 +1,7 @@
+.form-group {
+    &__legend {
+        // Bootstrap legend element's styles have 1.5rem font size
+        // for all legends on the page. Because of it we have to set size explicitly.
```
We probably want to override this globally if we don't like this style, I think we haven't used `<legend>` anywhere before so this was never noticed. There's probably a Sass variable to configure it

vovakulikov replied:
Not sure they have sass variable for it. 

Here is the snippet of legend styles from bootstrap code base

https://github.com/twbs/bootstrap/blob/7ea6578773cb1b7f5cfb8fb41321b3fa10349daf/scss/_reboot.scss#L501-L513

vovakulikov replied:
I've added `legend font-size: 1rem;` to global typography styles file

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/CreationSearchInsightPage.module.scss:
```diff
@@ -0,0 +1,5 @@
+.create-insight-page {
+    &__sub-title-container {
+        margin-bottom: 2rem;
```
Can this just use an `mb-*`? Would remove the need for this file :)

felixfbecker reviewed (commented):
Some more comments :)

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-series/FormSeries.module.scss:
```diff
@@ -0,0 +1,40 @@
+        margin-top: 0.5rem;
+    }
+
+    &__add-button {
+        padding: 1rem;
+        transition: background-color 0.2s ease, color 0.2s ease;
+
+        &:hover {
+            background-color: var(--link-hover-bg-color);
+        }
```
Can we instead rely on whatever `btn-secondary` provides?

vovakulikov replied:
Not sure. It seems more like link button and I add here some color transition for hover state but don't have any preference here. 

Secondary button 
<img width="823" alt="image" src="https://user-images.githubusercontent.com/18492575/116862601-aeac1e80-ac0d-11eb-8a64-4ef3692305c9.png">

Origin layout for this component from Figma 

<img width="489" alt="image" src="https://user-images.githubusercontent.com/18492575/116862820-ffbc1280-ac0d-11eb-9fbb-e8f83f960cf6.png">

AlicjaSuska replied:
Secondary button doesn't look good here. Let's use the text button as designed. We use analogical pattern for 'show more' on the home page:
<img width="815" alt="Screenshot 2021-05-03 at 20 22 26" src="https://user-images.githubusercontent.com/20326070/116916286-690f4600-ac4d-11eb-9466-75ea2f6267a8.png">

vovakulikov replied:
@AlicjaSuska I agree but what do you think if we add some background for hovered button just to to make 'visual' aiming easier


**Without state** 
<img width="575" alt="image" src="https://user-images.githubusercontent.com/18492575/116971391-276cb280-acc2-11eb-8120-341b489afa92.png">


**Hovered state**
<img width="672" alt="image" src="https://user-images.githubusercontent.com/18492575/116971333-158b0f80-acc2-11eb-8b1d-758ecf050dea.png">

felixfbecker replied:
If it's a link button, we have the `btn-link` button variant for that which should be used in that case.
Imo mixing link button style with a hover background that looks more like the secondary button makes it look more confusing/inconsistent, it's better to stick with the design system.

felixfbecker commented on client/web/src/insights/pages/search-insight-creation/components/form-series-input/FormSeriesInput.tsx:
```diff
@@ -0,0 +1,182 @@
+            />
+
+            <div className="mt-4">
+                <button type="submit" onClick={handleSubmitButton} className="btn btn-light">
```
This should be `btn-secondary` according to the design (`btn-light` we don't really use, it has very bad contrast and doesn't work well with light/dark theme)

```suggestion
                <button type="submit" onClick={handleSubmitButton} className="btn btn-secondary">
```

vovakulikov commented on client/web/src/insights/pages/search-insight-creation/components/creation-search-insight-form/CreationSearchInsightForm.module.scss:
```diff
@@ -0,0 +1,12 @@
+    &__separator {
+        margin-top: 2rem;
+        margin-bottom: 2rem;
+        width: 100%;
```
Since we use separator as flex child we have to set width explicitly here.

vovakulikov commented on client/web/src/insights/core/backend/requests/fetch-backend-insights.ts:
```diff
@@ -34,3 +39,19 @@ export function fetchBackendInsights(): Observable<InsightFields[]> {
     )
 }
+
+export function fetchLatestSubjectSettings(id: string): Observable<SubjectSettingsResult> {
```
I consider this as resolved just because It doesn't seem like a blocker but We should back to this problem later. Maybe migrate our API in FE from deprecated to new one gql handler.

vovakulikov commented:
@valerybugakov @eseliger @felixfbecker 
I've re-implemented form validation according to @felixfbecker comments above. I added native support validation logic but also I tried to stay with original API with hooks for form field management. This is a similar  approach compare to `useInputValidation` hook but extended because now in creation UI we need a bit flexible API. I believe we can align two mechanisms but first we have to agree what is the right for form fields validation and what is not. I created a [RFC about form validation](https://docs.google.com/document/d/1InTp9Kf0mtCPLM0g5yTgMEsSGZVlCKiuwPhQUpsKmts/edit?usp=sharing) if you got a time take a look (@AlicjaSuska )

Also I've added [story of creation ui page](https://www.chromatic.com/component?appId=5f0f381c0e50750022dc6bf7&name=web%2FCreationSearchInsightPage&specName=Page&buildNumber=18615) for better review.

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/creation-search-insight-form/CreationSearchInsightForm.tsx:
```diff
@@ -0,0 +1,253 @@
+
+import styles from './CreationSearchInsightForm.module.scss'
+
+const repositoriesFieldValidator = createRequiredValidator('Repositories is required field for code insight.')
```
```suggestion
const repositoriesFieldValidator = createRequiredValidator('Repositories is a required field.')
```

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/creation-search-insight-form/CreationSearchInsightForm.tsx:
```diff
@@ -0,0 +1,253 @@
+                .map(key => camelCase(key.split('.').pop()))
+        )
+
+        return composeValidators<string>(createRequiredValidator('Title is required field for code insight.'), value =>
```
```suggestion
        return composeValidators<string>(createRequiredValidator('Title is a required field.'), value =>
```

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/creation-search-insight-form/CreationSearchInsightForm.tsx:
```diff
@@ -0,0 +1,253 @@
+                title="Title"
+                autoFocus={true}
+                required={true}
+                description="Shown as title for your insight"
```
```suggestion
                description="Shown as the title for your insight"
```

Joelkw commented on client/web/src/insights/pages/search-insight-creation/CreationSearchInsightPage.tsx:
```diff
@@ -0,0 +1,139 @@
+                <h2>Create new code insight</h2>
+
+                <p className="text-muted">
+                    Search-based code insights analyse your code based on any search query.{' '}
```
```suggestion
                    Search-based code insights analyze your code based on any search query.{' '}
```

cc @AlicjaSuska I see you trying to sneak in your euro-centric spelling :)

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/creation-search-insight-form/CreationSearchInsightForm.tsx:
```diff
@@ -0,0 +1,253 @@
+                    name="visibility"
+                    value="personal"
+                    title="Personal"
+                    description="only for you"
```
```suggestion
                    description="only you"
```

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/creation-search-insight-form/CreationSearchInsightForm.tsx:
```diff
@@ -0,0 +1,253 @@
+                    name="visibility"
+                    value="organization"
+                    title="Organization"
+                    description="to all users in your organization"
```
```suggestion
                    description="all users in your organization"
```

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/creation-search-insight-form/CreationSearchInsightForm.tsx:
```diff
@@ -0,0 +1,253 @@
+            <FormGroup
+                name="visibility"
+                title="Visibility"
+                description="This insigh will be visible only on your personal dashboard. It will not be show to other
```
```suggestion
                description="This insight will be visible only on your personal dashboard. It will not be show to other
```

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/creation-search-insight-form/CreationSearchInsightForm.tsx:
```diff
@@ -0,0 +1,253 @@
+                name="visibility"
+                title="Visibility"
+                description="This insigh will be visible only on your personal dashboard. It will not be show to other
+                            users in your organisation."
```
```suggestion
                            users in your organization."
```

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/form-series-input/FormSeriesInput.tsx:
```diff
@@ -0,0 +1,114 @@
+                placeholder="Example: patternType:regexp const\s\w+:\s(React\.)?FunctionComponent"
+                description={
+                    <span>
+                        Do not include the <code>repo:</code> filter as it will be added automatically for the current
+                        repository
```
```suggestion
                        Do not include the <code>repo:</code> filter as it will be added automatically for the repositories you included above.
```

cc @AlicjaSuska I think this is a better description of the situation now?

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/form-series-input/FormSeriesInput.tsx:
```diff
@@ -0,0 +1,114 @@
+import { createRequiredValidator } from '../validators'
+
+const requiredNameField = createRequiredValidator('Name is required field for data series.')
+const validQuery = createRequiredValidator('Query is required field for data series.')
```
```suggestion
const validQuery = createRequiredValidator('Query is a required field for data series.')
```

vovakulikov replied:
@valerybugakov Agree we can come up with some predefined standard validators in the next releases when we have standard components or hooks for form validation, I definitely notice that. Thanks

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/form-series-input/FormSeriesInput.tsx:
```diff
@@ -0,0 +1,114 @@
+import { InputField } from '../form-field/FormField'
+import { createRequiredValidator } from '../validators'
+
+const requiredNameField = createRequiredValidator('Name is required field for data series.')
```
```suggestion
const requiredNameField = createRequiredValidator('Name is a required field for data series.')
```

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/form-series-input/FormSeriesInput.tsx:
```diff
@@ -0,0 +1,114 @@
+            />
+
+            <InputField
+                title="Query"
```
```suggestion
                title="Search query"
```

Wondering if we should explicitly call it a "search query" since query means other things in general graph data (sql query, graphlql query, etc) cc @AlicjaSuska

Joelkw commented on client/web/src/insights/pages/search-insight-creation/components/creation-search-insight-form/CreationSearchInsightForm.tsx:
```diff
@@ -0,0 +1,253 @@
+            <InputField
+                title="Repositories"
+                required={true}
+                description="Create a list of repositories to run your search over. Separate them with comas."
```
```suggestion
                description="Create a list of repositories to run your search over. Separate them with commas."
```

Joelkw reviewed (commented):
A couple of small copy notes in-line. 

One additional thought (maybe this is in one of the later PRs) – should we attach a max width to this form cc @AlicjaSuska ? On a large screen it gets pretty wiiiiiiiiiiiide :) 
![image](https://user-images.githubusercontent.com/11967660/117228220-80753d00-adcd-11eb-923a-aa992d8eb578.png)

vovakulikov commented:
@Joelkw Thanks for reviewing these labels. See this problem with the wide form on big screens. Probably this is my omission, I guess we should use width as code monitoring does with width container
`max-width: 66.6666666667%;`

valerybugakov reviewed (approved):
Huge work! Thanks, @vovakulikov, for addressing a lot of comments and putting extra effort into setting the foundation for form management. I totally agree with the UX aspects mentioned in [the form management RFC](https://docs.google.com/document/d/1InTp9Kf0mtCPLM0g5yTgMEsSGZVlCKiuwPhQUpsKmts/edit) and really like how they are implemented in the PR.

I left some comments, but they are optional and can be addressed in separate branches. It's always better to chunk work as much as possible, especially if the feature allows do it easily (it's possible to hide it under the development flag), and go with smaller, easier to review PRs. Keeping in mind that this the first iteration on the UI, and the functionality is under the flag not accessible to end-users, we can safely merge this PR ⛑. 

Posting a list of bullet points about the form management API that we discussed on the call to have them for future reference:

1. `formAPI` could be encapsulated inside of the `<Form />` component, which would then pass it down through the React context to the `<Field />` components.
2. It would allow avoiding rerenders of the whole form on each key-down in inputs. 
3. It would free developers from the burden of passing `valid`, `error`, `onChange`, etc., to input fields.
5. It's important to have _one_ way of doing things, like adding validators to the input fields.

## RE: Form Validation

In the current implementation, we support two ways of assigning validators to input fields: 

1. Native HTML attributes.
2. Custom validator functions.

The benefit of the first approach is that it exposes _the requirements and validity of an input field to assistive technologies automatically_. It's a great feature to have. But we still need custom validators because native attributes fall short on many validation rules that are usually applied to input fields. This issue is important, especially for non-trivial inputs like array fields, field values that depend on other form values, etc. The main takeaway here: we cannot use _only_ native HTML attributes for validation.

Custom validator functions cover all the use cases that can arise in forms because developers fully own this logic. Input requirements and validity of the input fields can be provided by aria attributes derived from the custom validators' information. This logic can be encapsulated into a component that will take care of it without bothering developers to handle such cases manually for every form.

We can support both ways to implement form validation, but having two ways of doing the same thing usually brings issues down the road and inconsistencies in the codebase. Also, it requires additional maintenance because editing this functionality requires keeping in mind _two_ ways of doing things and making additional effort to bridge these two worlds. Many open source solutions already solved most of the problems with the validation. We can use one of them as a drop-in solution. Thankfully to Vova's work, it should be easy because most of them use similar APIs. It would save us a lot of maintenance time and give one robust way of building forms with all the improvements I've mentioned above that we would need to implement ourselves. If we're not happy with any open-source solution out there, we should consider using one approach for validation in our in-house implementation. It will simplify the logic of the form management hooks and ensure that our forms are consistent across the board.

vovakulikov commented on client/web/src/insights/pages/search-insight-creation/components/creation-search-insight-form/CreationSearchInsightForm.test.tsx:
```diff
@@ -0,0 +1,250 @@
+    afterEach(cleanup)
+
+    const getFormFields = (getByRole: BoundFunction<GetByRole>) => {
+        const title = getByRole('textbox', { name: /title/i })
```
@valerybugakov Got it. I don't have preference here, but one concern about test-id is this isn't like users interact with UI. In tests we usually want to check users use cases so we repeat some actions like click and select some buttons but user can't see test-id they see inputs with name buttons wit text inside and etc.

vovakulikov commented on client/web/src/insights/pages/search-insight-creation/CreationSearchInsightPage.tsx:
```diff
@@ -0,0 +1,139 @@
+        history.push('/insights')
+    }, [history])
+
+    if (authenticatedUser === null) {
+        return <Redirect to="/" />
+    }
```
Nice catch. I left todo comment about this

felixfbecker replied:
That HOC even exists already :)
https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/blob/client/web/src/auth/withAuthenticatedUser.tsx#L10

vovakulikov commented on client/web/src/insights/pages/search-insight-creation/components/form-series/FormSeries.tsx:
```diff
@@ -0,0 +1,152 @@
+
+    // Add empty series form component that user be able to fill this form and create
+    // another chart series.
+    const handleAddSeries = useCallback(() => {
```
@valerybugakov logic for deleting series this is what we want to add in next version of creation UI. I agree it should be here and we have a design for it already. So I will add this logic in next PRs. 

About useCallbacks I guess I can easily remove them - we have some components below where we're passing them but these component don't have useMemo so it does't change anything if we have here pure callback or useCallback. Thanks for catching.

vovakulikov replied:
Thing that I've noticed usually this rule `react/jsx-no-bind ` doesn't make sense because we don't use `useMemo` on react function components.

vovakulikov commented:
@valerybugakov Huge thanks for the comprehensive comment about validation and for the review itself. 
I closed almost all issues and comments in this PR (I left some minor things which would be fixed in the next releases of creation UI, and now these changes under local storage). I'm about to merge this PR as soon as all PR checks pass. Thank you to all those who were involved!

felixfbecker reviewed (commented):
Awesome work @vovakulikov ! 👏🏻
